File: apps/api/schemas/decision_v2.py
"""
Decision API v2 Schemas & Data Contracts (Phase 6)
"""
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from backend_worker.ranking_v2.master_rank_v2 import ThesisBand, FactorDriver
from backend_worker.setup.setup_engine import SetupState
from backend_worker.risk.risk_engine import RiskState, DataGrade

class PriceQuote(BaseModel):
    value: float
    currency: str
    change_pct: float
    as_of: str

class MasterRankDetails(BaseModel):
    score: float
    band: ThesisBand
    segment_percentile: float
    weighted_coverage: float
    model_version: str

class SetupDetails(BaseModel):
    state: SetupState
    ui_label_sv: str
    reason_codes: List[str] = Field(default_factory=list)

class RiskDetails(BaseModel):
    state: RiskState
    dominant_risk: str
    liquidity_grade: str
    risk_flags: List[str] = Field(default_factory=list)

class DataGradeDetails(BaseModel):
    grade: DataGrade
    weighted_coverage: float
    critical_warnings: List[str] = Field(default_factory=list)

class DecisionRowV2(BaseModel):
    decision_snapshot_id: str
    listing_id: str
    ticker: str
    name: str
    segment: str
    sector: Optional[str] = None
    country: str = "SE"
    price: PriceQuote
    master_rank: MasterRankDetails
    setup: SetupDetails
    risk: RiskDetails
    data_grade: DataGradeDetails
    positive_drivers: List[FactorDriver] = Field(default_factory=list)
    negative_drivers: List[FactorDriver] = Field(default_factory=list)

class ScreenerResponseV2(BaseModel):
    total_count: int
    rows: List[DecisionRowV2]
    as_of: str
    snapshot_id: str
    active_filters: Dict[str, Any] = Field(default_factory=dict)

class StockDecisionV2(BaseModel):
    decision_snapshot_id: str
    listing_id: str
    ticker: str
    name: str
    segment: str
    sector: Optional[str] = None
    country: str
    price: PriceQuote
    master_rank: MasterRankDetails
    setup: SetupDetails
    risk: RiskDetails
    data_grade: DataGradeDetails
    positive_drivers: List[FactorDriver] = Field(default_factory=list)
    negative_drivers: List[FactorDriver] = Field(default_factory=list)
    factor_scores: Dict[str, Optional[float]] = Field(default_factory=dict)
    factor_reliabilities: Dict[str, float] = Field(default_factory=dict)
    warnings: List[str] = Field(default_factory=list)

File: apps/web/lib/types/decision_v2.ts
/**
 * MarketScan Decision API v2 TypeScript Contracts (Phase 6)
 * Single source of truth for Screener, Stock Page, Compare, Watchlist, and Portfolio.
 */

export type ThesisBand =
  | "EXCEPTIONAL"
  | "STRONG"
  | "POSITIVE"
  | "MIXED"
  | "WEAK"
  | "INSUFFICIENT";

export type SetupState =
  | "CONFIRMED"
  | "PULLBACK"
  | "NEUTRAL"
  | "EXTENDED"
  | "DAMAGED"
  | "EVENT_RISK"
  | "INSUFFICIENT";

export type RiskState =
  | "LOW"
  | "MEDIUM"
  | "HIGH"
  | "VERY_HIGH"
  | "EVENT"
  | "INSUFFICIENT";

export type DataGrade = "A" | "B" | "C" | "D" | "E" | "F";

export interface FactorDriver {
  factor_name: string;
  label_sv: string;
  raw_score: number | null;
  reliability: number;
  contribution: number;
}

export interface PriceQuote {
  value: number;
  currency: string;
  change_pct: number;
  as_of: string;
}

export interface MasterRankDetails {
  score: number;
  band: ThesisBand;
  segment_percentile: number;
  weighted_coverage: number;
  model_version: string;
}

export interface SetupDetails {
  state: SetupState;
  ui_label_sv: string;
  reason_codes: string[];
}

export interface RiskDetails {
  state: RiskState;
  dominant_risk: string;
  liquidity_grade: string;
  risk_flags: string[];
}

export interface DataGradeDetails {
  grade: DataGrade;
  weighted_coverage: number;
  critical_warnings: string[];
}

export interface DecisionRowV2 {
  decision_snapshot_id: string;
  listing_id: string;
  ticker: string;
  name: string;
  segment: string;
  sector?: string | null;
  country: string;
  price: PriceQuote;
  master_rank: MasterRankDetails;
  setup: SetupDetails;
  risk: RiskDetails;
  data_grade: DataGradeDetails;
  positive_drivers: FactorDriver[];
  negative_drivers: FactorDriver[];
}

export interface ScreenerResponseV2 {
  total_count: number;
  rows: DecisionRowV2[];
  as_of: string;
  snapshot_id: string;
  active_filters: Record<string, unknown>;
}

export interface StockDecisionV2 extends DecisionRowV2 {
  factor_scores: Record<string, number | null>;
  factor_reliabilities: Record<string, number>;
  warnings: string[];
}
